### ISSUE TITLE
Auth vs auth - http-hooks and roadmap

### ISSUE THREAD

**Issue by ozra** [erlio/vernemq#629]
We currently use http-hooks for authentication and authorization, after realizing the lua-plugin was an inefficient way. We have some worries about the MQ in relation to hooks-usage.

### Setup:
- Clients initially connect with a UUID/v4, in base62 22 chars, as clientId.
- Connect-level authentication is basically a no-op
- Those that have a login, and do login, authenticate app-level via API via MQTT - these topics are allowed for "guests" (which are all clientId's not having an authenticated session)
- This creates a session-entry with their login's grants
- The session-id is - as everything here - (UUID/v4).b62
    - A message is sent to the client with this if auth was ok
    - They disconnect and reconnect with this as clientId
- Every API-endpoint already has clientId as part of topic
- Right now all authorization works via http-hooks
- Our ACL-model would map pretty much 1:1 to the structure VerneMQ favours, as per the DB-auth methods.

### Problem (maybe not?):
- If we don't send cache-headers, VerneMQ will do http-requests for each single pub/sub - inefficient
    - our services are many, our entire system is mqtt-API-based - client<=>service, service<=>service, a lot of messaging going on
- If we don't, it will cache indefinitely, according to documentation
    - When users end a session (which may happen rarely on Mobile-devices, but more often from web-clients), that clientId is out of the world for ever. Those cached entries are RAM-hogs. The user will cause new cached entries on next new session. RAM BOOM!

### The question on VMQ-webhooks really:
- Could http-hooks be configured with an endpoint that returns an ACL-json — like the DB-variants? Instead of topic for topic
- We also use no part of payload for auth, but it seems impossible to configure away as part of hook-request? Also couldn't figure out how to at least remove base64-enc.
- Could VMQ-hooks also ditch old cached entries properly based on cache-headers? Avoiding RAM-explosion?
- _More shortly put_: how prioritized is the http-hooks module in VerneMQ roadmap? _Are we going down a bad road here? Should we consider other MQ options?_

The MQTT-broker is the only "single point of failure section" in our stack — the rest is very decentralized, so we'd really like to come up with a solution that doesn't require restarting VMQ unnecessarily.  Sorry if this is the wrong place for these questions on VMQ/web-hooks/roadmap!

**Comment by ioolkos** [comment]
@ozra you're very welcome to ask any kind of questions here. (but also welcome to contact us per mail anytime).
I can't go through the setup and points right now, but we'll have a close look at all of them. 
I don't see any show stoppers to really worry about, in short.

**Comment by ozra** [comment]
I guess it's just the last actual questions that is important when tl;dr :)

**Comment by larshesel** [comment]
Hi @ozra 

The plugin hasn't had a lot of attention lately. The points you bring up about the cache using a lot of memory (not being garbage collected) as well as the payload being sent on the `auth_on_publish` hook are known limitations - but we haven't prioritized dealing with them as there hasn't been any real demand. So at the moment it doesn't have so high priority - but we of course are always open to accept pull request or sponsoring. Which of the features is the most important one for your use case?

I think the idea of having webhooks work a bit like the DB auth plugins is a good one, but I also think that would be a major change to how the current plugin works. Maybe something like that should be it's own plugin. 

About the base64 encoding this can be disabled when registering the hooks - see `vmq-admin webhooks register --help`. Just note that if the payload has to be base64 encoded as json cannot encode arbitrary binary data.

**Comment by ozra** [comment]
Oh. Well really any way where app-logic can react to the authentication and prepare acl entries for the authorization. We first did lua, but realized it was a slow option. It sounds reasonable to make it its' own instead of modifying the existent one. Regarding _64, we've got everything in the conf, we figure that's where ot should be set?

**Comment by ioolkos** [comment]
@ertanden thanks for your comment. 
Both of your points make sense to me. @larshesel @dergraf reopen issue and reopen discussion?

**Comment by codeadict** [comment]
I also think the idea of having webhooks work a bit like the DB auth plugins kind of defeats the purpose of the webhooks plugin converting it in a do it all or maybe it can be configured in an elegant way to do it? The bit about garbage collecting is something that can be tasked out and I think would be a great improvement, I developed a custom plugin that also uses similar cache and had the issue until I implemented GC. GC the cache will be something that I may be able to work on if I get time but can't promise, have many things hanging and short in time.
